src/iop/prover.rs

```diff
@@ -0,0 +1,821 @@
+        index: &'a Index<F>,
+        c: C,
+    ) -> Result<ProverState<'a, F>, Error> {
+        let init_time = start_timer!(|| "IOP::Prover::Init");
+
+        let witnesses_time = start_timer!(|| "Compute witnesses");
+        let mode = SynthesisMode::Prove {
+            construct_matrices: false,
+        };
+        let mut pcs = ConstraintSystem::new(mode);
+        c.generate_constraints(&mut pcs)?;
+        end_timer!(witnesses_time);
+
+        let num_non_zero = index.index_info.num_non_zero;
+
+        let ConstraintSystem {
+            input_assignment: formatted_input_assignment,
+            aux_assignment: witness_assignment,
+            num_constraints,
+            ..
+        } = pcs;
+
+        let num_input_variables = formatted_input_assignment.len();
+        let num_witness_variables = witness_assignment.len();
+        if index.index_info.num_constraints != num_constraints
+            || num_input_variables != index.index_info.num_inputs
+            || num_witness_variables != index.index_info.num_witness
+        {
+            return Err(Error::InstanceDoesNotMatchIndex);
+        }
+
+        let (domain_h, domain_k, domain_x, domain_b) = Self::build_domains(
+            num_witness_variables,
+            num_input_variables,
+            num_constraints,
+            num_non_zero,
+        )?;
+
+        end_timer!(init_time);
+
+        Ok(ProverState {
+            formatted_input_assignment,
+            witness_assignment,
+            w_poly: None,
+            my_polys: None,
+            index,
+            verifier_first_msg: None,
+            domain_h,
+            domain_k,
+            domain_x,
+            domain_b,
+        })
+    }
+
+    /// Prover first round of the algebraic oracle proof, the initial round in [HGB].
+    /// Determines the oracles for the witness-related polynomials
+    /// `w(X)`, `y_A(X)` and `y_B(X)`.
+    ///
+    /// [HGB]: https://eprint.iacr.org/2021/930
+    pub fn prover_first_round<'a, R: RngCore>(
+        mut state: ProverState<'a, F>,
+        zk: bool,
+        rng: &mut R,
+    ) -> Result<(ProverFirstOracles<F>, ProverState<'a, F>), Error> {
+        let round_time = start_timer!(|| "IOP::Prover::FirstRound");
+        let domain_h = &state.domain_h;
+
+        let x_time = start_timer!(|| "Computing x polynomial and evals");
+        let domain_x = &state.domain_x;
+        let x_poly = EvaluationsOnDomain::from_vec_and_domain(
+            state.formatted_input_assignment.clone(),
+            domain_x.clone(),
+        )
+        .interpolate();
+        // Evaluate the input polynomial x(X) over H.
+        let x_evals = domain_h.fft(&x_poly);
+        end_timer!(x_time);
+
+        /* Compute the normalized witness polynomial w(X) which allows easy
+        combination with the input polynomial,
+            y(X) = x(X) + z_I(X)* w(X) mod (X^n-1).
+        */
+        let ratio = domain_h.size() / domain_x.size();
```

Comment:
We can use the get_subdomain_step function ?

Reply:
Done